src/retry-interceptor.js

```diff
@@ -0,0 +1,111 @@
+import { Interceptor, RetryConfiguration } from './interfaces';
+
+export const retryStrategy = {
+  fixed: 0,
+  incremental: 1,
+  exponential: 2,
+  random: 3
+};
+
+const defaultRetryConfig: RetryConfiguration = {
+  maxRetries: 3,
+  interval: 1000,
+  strategy: retryStrategy.fixed
+};
+
+export class RetryInterceptor implements Interceptor {
+  retryConfig: RetryConfiguration;
+
+  constructor(retryConfig?: RetryConfiguration) {
+    this.retryConfig = Object.assign({}, defaultRetryConfig, retryConfig || {});
+
+    if (this.retryConfig.strategy === retryStrategy.exponential &&
+       this.retryConfig.interval <= 1000) {
+      throw new Error('An interval less than or equal to 1 second is not allowed when using the exponential retry strategy');
+    }
+  }
+
+  request(request) {
+    if (!request.retryConfig) {
+      request.retryConfig = Object.assign({}, this.retryConfig);
+      request.retryConfig.counter = 0;
+    }
+    return request;
+  }
+
+  response(response, request) {
+    // retry was successful, so clean up after ourselves
+    delete request.retryConfig;
+    return response;
+  }
+
+  responseError(error, request) {
+    const { retryConfig } = request;
+    const { requestClone } = retryConfig;
+    return Promise.resolve().then(() => {
+      if (retryConfig.counter < retryConfig.maxRetries) {
+        const result = retryConfig.doRetry ? retryConfig.doRetry(error, request) : true;
+
+        return Promise.resolve(result).then(doRetry => {
+          if (doRetry) {
+            retryConfig.counter++;
+
+            return new Promise(resolve => setTimeout(resolve, calculateDelay(retryConfig) || 0)).then(() => {
+              let newRequest = requestClone.clone();
+              if (typeof (retryConfig.beforeRetry) === 'function') {
+                return retryConfig.beforeRetry(newRequest, retryConfig.client);
+              }
+              return newRequest;
+            }).then(newRequest => {
+              return retryConfig.client.fetch(Object.assign(newRequest, { retryConfig }));
+            });
```

Comment:
Not needed? The interceptor can just return a `Request`, and the HttpClient will start a new fetch https://github.com/aurelia/fetch-client/blob/master/src/http-client.js#L116

3 new Promises are being created. Are they all needed, or can we accomplish the same with less? I haven't checked up on this.

Reply:
Dude, that's really cool. I didn't know that.

Reply:
I'm not sure your correct about that. I'll test it to see though, as it would be nice if you are. Even if this is correct, I think users would likely expect the full http client "build a request from my defaults" process to run, especially on aretry where they possibly have updated the default headers to update an expired auth token.

Reply:
Really tearing in to @CasiOo there.

Reply:
Ah too bad :( 
Is there any other way we can get around `HttpClient` knowing about retrying, and for the interceptor to replay a request?
Developers implementing custom interceptors might want to replay/start a new request just like we're doing here. Can they do it without having to modify the HttpClient code itself?

Reply:
Not a personal attack on @CasiOo at all. :-)

>Developers implementing custom interceptors might want to replay/start a new request just like we're doing here. Can they do it without having to modify the HttpClient code itself?

Well, with this change, technically they can simply by putting a non-falsey object property on the request called `retryConfig` in their own interceptor's `request` callback (since that will tell the HttpClient to put the request clone and itself on that property). I'm with you on not wanting to have retry-specific code in the HttpClient, but at the same time, I question if it is a good idea to always give developers the `HttpClient` instance, as I feel like that is likely to encourage developers to do things that might not be advisable. Given that, I'm in favor of leaving the code as is. I think that, since we would be shipping the `RetryInterceptor` implementation "in the box" with the client, I think it's okay to have the couple of lines of `RetryInterceptor` specific code in the client itself.